actors/builtin/storage_power/storage_power_actor_state.go

```diff
@@ -91,28 +102,35 @@ func addrInArray(a addr.Address, list []addr.Address) bool {
 }
 
 // _selectMinersToSurprise implements the PoSt-Surprise sampling algorithm
-func (st *StoragePowerActorState) _selectMinersToSurprise(challengeCount int, randomness abi.Randomness) []addr.Address {
+func (st *StoragePowerActorState) _selectMinersToSurprise(rt vmr.Runtime, challengeCount int, randomness abi.Randomness) []addr.Address {
 	// this wont quite work -- a.PowerTable is a HAMT by actor address, doesn't
 	// support enumerating by int index. maybe we need that as an interface too,
 	// or something similar to an iterator (or iterator over the keys)
 	// or even a seeded random call directly in the HAMT: myhamt.GetRandomElement(seed []byte, idx int) using the ticket as a seed
 
-	ptSize := len(st.PowerTable)
-	allMiners := make([]addr.Address, len(st.PowerTable))
-	index := 0
+	// Reply to above comment: for now do the easy thing and use the hamt ForEach method as lotus does
 
-	for address := range st.PowerTable {
-		allMiners[index] = address
+	var index int64
+	var allMiners []addr.Address
+	if err := adt.NewMap(vmr.AsStore(rt), st.PowerTable).ForEach(func(k string, v interface{}) error {
+		maddr, err := addr.NewFromBytes([]byte(k))
+		if err != nil {
+			return err
+		}
+		allMiners[index] = maddr
```

Comment:
Index out of bounds. But also the index seems completely unused, remove it and append to slice instead.